src/cdomains/apron/octApronDomain.apron.ml

```diff
@@ -56,504 +314,466 @@ struct
+      if exp_is_cons e then
+        e
+      else
+        (* convert non-constraint expression, such that we assert(e != 0) *)
+        BinOp (Ne, e, zero, intType)
+    in
+    assert_cons d x negate
+
+  let check_assert d e =
+    if is_bot_env (assert_inv d e false) then
+      `False
+    else if is_bot_env (assert_inv d e true) then
+      `True
+    else
+      `Top
+
+  let int_of_scalar (scalar: Scalar.t) =
+    if Scalar.is_infty scalar <> 0 then (* infinity means unbounded *)
+      None
+    else
+      match scalar with
+      | Mpqf scalar when Mpzf.cmp_int (Mpqf.get_den scalar) 1 = 0 -> (* rational must be integer (denominator 1) *)
+        Some (IntOps.BigIntOps.of_string (Mpqf.to_string scalar))
+      | _ ->
+        failwith "int_of_scalar: not integer"
+
+  (** Evaluate non-constraint expression as interval. *)
+  let eval_interval_expr d e =
+    match Convert.texpr1_of_cil_exp (A.env d) e with
+    | texpr1 ->
+      let bounds = A.bound_texpr Man.mgr d texpr1 in
+      let min = int_of_scalar bounds.inf in
+      let max = int_of_scalar bounds.sup in
+      (min, max)
+    | exception Convert.Unsupported_CilExp ->
+      (None, None)
+
+  (** Evaluate constraint or non-constraint expression as integer. *)
+  let eval_int d e =
+    let module ID = Queries.ID in
+    let ik = Cilfacade.get_ikind_exp e in
+    if exp_is_cons e then
+      match check_assert d e with
+      | `True -> ID.of_bool ik true
+      | `False -> ID.of_bool ik false
+      | `Top -> ID.top ()
+    else
+      match eval_interval_expr d e with
+      | (Some min, Some max) -> ID.of_interval ik (min, max)
+      | (Some min, None) -> ID.starting ik min
+      | (None, Some max) -> ID.ending ik max
+      | (None, None) -> ID.top ()
+
+
+  let assign_var_handling_underflow_overflow oct v e =
+    let ikind = Cilfacade.get_ikind v.vtype in
+    let signed = Cil.isSigned ikind in
+    let new_oct = assign_exp oct (Var.of_string v.vname) e in
+    let lower_limit, upper_limit = IntDomain.Size.range_big_int ikind in
+    let check_max =
+      check_assert new_oct (BinOp (Le, Lval (Cil.var @@ v), (Cil.kintegerCilint ikind (Cilint.cilint_of_big_int upper_limit)), intType)) in
+    let check_min =
+      check_assert new_oct (BinOp (Ge, Lval (Cil.var @@ v), (Cil.kintegerCilint ikind (Cilint.cilint_of_big_int lower_limit)), intType)) in
+    if signed then
+      if check_max <> `True || check_min <> `True then
+        if GobConfig.get_bool "ana.octapron.no_signed_overflow" then
+          new_oct
+        else
+          (* Signed overflows are undefined behavior, so octagon goes to top if it might have happened. *)
+          top_env (A.env oct)
+      else
+        new_oct
+    else if check_max <> `True || check_min <> `True then
+      (* Unsigned overflows are defined, but for now
+          the variable in question goes to top if there is a possibility of overflow. *)
+      forget_vars oct [Var.of_string v.vname]
+    else
+      new_oct
```

Comment:
I'm really skeptical about this function because of https://github.com/goblint/analyzer/issues/273#issuecomment-875369259. I'm not sure it's worth handling overflows in octApron at all if it's incomplete like that.

Its signature also doesn't fit the rest because it's the only function here that requires a `varinfo` instead of `Apron.Var.t`. So I've really wanted to refactor this but I'm unsure what to do about this overflow stuff.